modules/percolator/src/main/java/org/elasticsearch/percolator/ExtractQueryTermsService.java

```diff
@@ -104,24 +118,30 @@ public static void extractQueryTerms(Query query, ParseContext.Document document
      * If from part of the query, no query terms can be extracted then term extraction is stopped and
      * an UnsupportedQueryException is thrown.
      */
-    static Set<Term> extractQueryTerms(Query query) {
-        if (query instanceof MatchNoDocsQuery) {
-            // no terms to extract as this query matches no docs
-            return Collections.emptySet();
+    // TODO: restructure this method? For example have a Function<Query, Result> per supported query.
+    // All functions then sit in a Map<Class<? extends Query>, Function<Query, Result>> and we lookup via map.get(query.getClass())
+    static Result extractQueryTerms(Query query) {
+        if (query instanceof MatchAllDocsQuery) {
+            // we need to emit an empty term otherwise a percolator query with this query may never end up as a
+            // candidate match:
+            return new Result(true, new Term("", ""));
```

Comment:
this hack looks a bit dangerous to me, maybe we need a special extraction status for it rather than using the empty term?